cpp/src/reductions/segmented/simple.cuh

```diff
@@ -214,23 +212,57 @@ std::unique_ptr<column> fixed_point_segmented_reduction(
   rmm::mr::device_memory_resource* mr)
 {
   using RepType = device_storage_type_t<InputType>;
-  return simple_segmented_reduction<RepType, RepType, Op>(
-    col, offsets, null_handling, init, stream, mr);
-}
+  auto result =
+    simple_segmented_reduction<RepType, RepType, Op>(col, offsets, null_handling, init, stream, mr);
+  auto const scale = [&] {
+    if constexpr (std::is_same_v<Op, cudf::reduction::op::product>) {
+      // The product aggregation requires updating the scale of the fixed-point output column.
+      // The output scale needs to be the maximum valid count of all segments multiplied by
+      // the input scale value.
+      auto d_col = column_device_view::create(col, stream);
+      rmm::device_uvector<size_type> valid_counts =
+        cudf::reduction::detail::segmented_valid_counts(*d_col,
+                                                        col.has_nulls(),
+                                                        offsets,
+                                                        null_handling,
+                                                        stream,
+                                                        rmm::mr::get_current_device_resource());
+
+      auto max_valid_count = thrust::reduce(rmm::exec_policy(stream),
```

Comment:
I think it's a bit more expressive to use `thrust::max_element` here directly

Reply:
I disagree. The `max_element` does not return an answer but a pointer/iterator to the element. I prefer `reduce` do that step for me.

Reply:
That's fair, since it looks like the result is needed on host right afterwards